src/dscope.d

```diff
@@ -425,25 +442,84 @@ struct Scope
             }
             return null;
         }
-        for (Scope* sc = &this; sc; sc = sc.enclosing)
+
+        Dsymbol searchScopes(int flags)
         {
-            assert(sc != sc.enclosing);
-            if (!sc.scopesym)
-                continue;
-            //printf("\tlooking in scopesym '%s', kind = '%s'\n", sc->scopesym->toChars(), sc->scopesym->kind());
-            if (Dsymbol s = sc.scopesym.search(loc, ident, flags))
+            for (Scope* sc = &this; sc; sc = sc.enclosing)
             {
-                if (ident == Id.length && sc.scopesym.isArrayScopeSymbol() && sc.enclosing && sc.enclosing.search(loc, ident, null, flags))
+                assert(sc != sc.enclosing);
+                if (!sc.scopesym)
+                    continue;
+                //printf("\tlooking in scopesym '%s', kind = '%s'\n", sc->scopesym->toChars(), sc->scopesym->kind());
+                if (Dsymbol s = sc.scopesym.search(loc, ident, flags))
                 {
-                    warning(s.loc, "array 'length' hides other 'length' name in outer scope");
+                    if (!(flags & IgnoreThisModule) &&
+                        ident == Id.length && sc.scopesym.isArrayScopeSymbol() &&
+                        sc.enclosing && sc.enclosing.search(loc, ident, null, flags))
+                    {
+                        warning(s.loc, "array 'length' hides other 'length' name in outer scope");
+                    }
+                    //printf("\tfound local %s.%s, kind = '%s'\n", s.parent ? s.parent.toChars() : "", s.toChars(), s.kind());
+                    if (pscopesym)
+                        *pscopesym = sc.scopesym;
+                    return s;
                 }
-                //printf("\tfound %s.%s, kind = '%s'\n", s->parent ? s->parent->toChars() : "", s->toChars(), s->kind());
-                if (pscopesym)
-                    *pscopesym = sc.scopesym;
-                return s;
+                if (sc.scopesym.isModule())
+                    break;
             }
+            return null;
         }
-        return null;
+
+        version (none)
+        {
+            Dsymbol sx = searchScopes(flags);       // look in both locals and imports
+            version (LOGSEARCH)
+            {
+                if (sx)
+                    printf("*Scope.search() found  %s.%s, kind = '%s'\n", sx.parent ? sx.parent.toChars() : "", sx.toChars(), sx.kind());
+                else
+                    printf("*Scope.search() not found\n");
+            }
+        }
+
+        // First look in local scopes
+        Dsymbol s = searchScopes(flags | IgnoreImports);
+
+        version (none)
+            if (s && sx != s)
+            {
```

Comment:
btw one nice thing:

```d
version(whatevs) if (s && sx != x)
{
}
```

i.e. putting the two on the same line - clarifies that version and if are one item. So no need for two levels of indentation either.